### PR TITLE
New merge master 2.9.99

### PR DESCRIPTION
**Pull request by ndellingwood**
Master branch updated with kokkos 2.9.99, this PR merges master back into develop.

### CODE REVIEW

**Comment by ndellingwood** [comment]
This replaces #2718. Merge into master performed after merge of #2688. All merge conflicts were resolved by deferring to the develop branch.

**Comment by jjwilke** [comment]
Do people want me to squash these? There's really not that much net change for 42 commits?

File: BUILD.md [hunk 1]
```diff
@@ -157,9 +157,6 @@ Options can be enabled by specifiying `-DKokkos_ENABLE_X`.
 * Kokkos_ENABLE_DEPRECATED_CODE
     * Whether to enable deprecated code
     * BOOL Default: OFF
```
Comment (jjwilke) [inline BUILD.md]: I think we want examples now, yeah?

File: CMakeLists.txt [hunk 2]
```diff
@@ -120,14 +120,18 @@ ENDIF()
...
 # but scoping issues can make it difficult
-GLOBAL_SET(KOKKOS_COMPILE_OPTIONS)
-GLOBAL_SET(KOKKOS_LINK_OPTIONS)
-GLOBAL_SET(KOKKOS_CUDA_OPTIONS)
-GLOBAL_SET(KOKKOS_CUDAFE_OPTIONS)
-GLOBAL_SET(KOKKOS_XCOMPILER_OPTIONS)
+GLOBAL_RESET(KOKKOS_COMPILE_OPTIONS)
+GLOBAL_RESET(KOKKOS_LINK_OPTIONS)
+GLOBAL_RESET(KOKKOS_CUDA_OPTIONS)
+GLOBAL_RESET(KOKKOS_CUDAFE_OPTIONS)
```
Comment (jjwilke) [inline CMakeLists.txt]: My unnecessary changes creeped back in.

File: cm_generate_makefile.bash [hunk 3]
```diff
@@ -343,5 +335,5 @@ if [[ ${COMPILER} == *clang* ]]; then
    fi 
 fi
  
-echo cmake $COMPILER_CMD  -DCMAKE_CXX_FLAGS="${KOKKOS_CXXFLAGS}" -DCMAKE_EXE_LINKER_FLAGS="${KOKKOS_LDFLAGS}" -DCMAKE_INSTALL_PREFIX=${PREFIX} ${KOKKOS_DEVICE_CMD} ${KOKKOS_ARCH_CMD} -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=${KOKKOS_DO_EXAMPLES} ${KOKKOS_OPTION_CMD} ${KOKKOS_CUDA_OPTION_CMD} -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF ${STANDARD_CMD} ${KOKKOS_DEBUG_CMD} ${KOKKOS_PATH}
-cmake $COMPILER_CMD  -DCMAKE_CXX_FLAGS="${KOKKOS_CXXFLAGS//\"}" -DCMAKE_EXE_LINKER_FLAGS="${KOKKOS_LDFLAGS//\"}" -DCMAKE_INSTALL_PREFIX=${PREFIX} ${KOKKOS_DEVICE_CMD} ${KOKKOS_ARCH_CMD} -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=${KOKKOS_DO_EXAMPLES} ${KOKKOS_OPTION_CMD} ${KOKKOS_CUDA_OPTION_CMD} -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF ${STANDARD_CMD} ${KOKKOS_DEBUG_CMD} ${KOKKOS_PATH}
+echo cmake $COMPILER_CMD  -DCMAKE_CXX_FLAGS="${KOKKOS_CXXFLAGS}" -DCMAKE_EXE_LINKER_FLAGS="${KOKKOS_LDFLAGS}" -DCMAKE_INSTALL_PREFIX=${PREFIX} ${KOKKOS_DEVICE_CMD} ${KOKKOS_ARCH_CMD} -DKokkos_ENABLE_TESTS=ON ${KOKKOS_OPTION_CMD} ${KOKKOS_CUDA_OPTION_CMD} -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF ${STANDARD_CMD} ${KOKKOS_DEBUG_CMD} ${KOKKOS_PATH}
```
Comment (jjwilke) [inline cm_generate_makefile.bash]: I think we do want examples, yeah?

File: cmake/fake_tribits.cmake [hunk 4]
```diff
@@ -24,6 +24,10 @@ IF(NOT ${MACRO_DEFINE_NAME} STREQUAL "")
 ENDIF()
 ENDMACRO()
 
+MACRO(GLOBAL_RESET VARNAME)
+  SET(${VARNAME} "" CACHE INTERNAL "" FORCE)
```
Comment (jjwilke) [inline cmake/fake_tribits.cmake]: Unnecessary changes creeped back in.

**Comment by dalg24** [comment]
Again I see no compelling reason to merge anything else than the update the CHANGELOG and `master_history.txt`.

I saw https://github.com/kokkos/kokkos/pull/2718#issuecomment-582059159 but I'd argue this is outdated.  We clearly state in https://github.com/kokkos/kokkos/blob/master/CONTRIBUTING.md#pull-requests that branches should be created from (or rebased onto) `develop`.

**Comment by crtrott** [comment]
Yes only CHANGELOG and master_history need to go back in.

**Comment by ndellingwood** [comment]
@crtrott @dalg24 I'll create a new PR with just the CHANGELOG and master_history merged from master over to develop.

**Comment by ndellingwood** [comment]
Replaced by #2730, closing.

**Comment by dalg24** [comment]
> @crtrott @dalg24 I'll create a new PR with just the CHANGELOG and master_history merged from master over to develop.

Next time please just push force instead of opening a new one.
